site/public/js/gradeable-notebook.js

```diff
@@ -65,6 +65,161 @@ function setCodeBox(codebox_id, state)
     }
 }
 
+const AUTOSAVE_KEY = `${window.location.pathname}-autosave`;
+
+/**
+ * This value will be true if autosave is enabled for this session.
+ * 
+ * Specifically, we try to set a test value in localStorage and immediately 
+ * remove the test value. If all goes well, then localStorage is available and
+ * we can auto-save to there. Otherwise, we simply disable auto-saving.
+ * 
+ * In particular, setting an item can fail in one of two scenarios:
+ * * The local storage quota is full
+ * * The `localStorage` variable simply doesn't exist and therefore the browser doesn't support it.
+ * 
+ * The first point isn't too relevant as we don't expect to be storing a ton of
+ * data; however, Safari on iOS in Private Mode sets the quota to zero, meaning 
+ * that every attempt to write to localStorage will throw an exception. 
+ * Throwing an exception every keypress isn't great, hence the primary 
+ * motivation for this block here.
+ * 
+ * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API#Feature-detecting_localStorage|MDN} on detecting localStorage.
+ * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Storage/setItem|Storage.setItem()} documentation.
+ */
+const autosaveEnabled = (() => {
+    try {
+        localStorage.setItem('TEST', 'TEST');
+        localStorage.removeItem('TEST');
+        return true;
+    } catch (e) {
+        return false;
+    }
+})();
+
+function msToDays(ms) {
+    return ms / (1000 * 60 * 60 * 24);
+}
+
+/**
+ * Scans all autosaved data and deletes any entries older than thirty days.
+ */
+function cleanupAutosaveHistory() {
+    if (autosaveEnabled) {
+        const toDelete = [];
+        for (let i = 0; i < localStorage.length; i++) {
+            const key = localStorage.key(i);
+            try {
+                const { type, timestamp } = JSON.parse(localStorage.getItem(key));
+                
+                // Skip over non-autosave entries.
+                if (type !== "notebook-autosave") {
+                    continue;
+                }
+
+                if (msToDays(Date.now() - timestamp) > 30) {
+                    toDelete.push(key);
+                }
+            } catch (e) {
+                // If this item in localStorage isn't a JSON blob (possible as
+                // we use localStorage in a couple of other things), then
+                // JSON.parse() will throw a SyntaxError. We don't care about
+                // these entries, so skip over them.
+                if (e instanceof SyntaxError) {
+                    continue;
+                } else {
+                    throw e;
+                }
+            }
+        }
+        toDelete.forEach(localStorage.removeItem, localStorage);
```

Comment:
Let's use the [endswith](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith) to limit the keys we care about (just need unique suffix) instead of parsing all keys. This has the added benefit of that if any value in localStorage related to this feature gets corrupted for whatever reason, we can properly just dispose of it instead of it hanging around infinitely.
```suggestion
        for (let i = 0; i < localStorage.length; i++) {
            const key = localStorage.key(i);
            if (!key.endswith('-notebook-autosave')) {
                continue
            }
            try {
                const { timestamp } = JSON.parse(localStorage.getItem(key));

                if (msToDays(Date.now() - timestamp) > 30) {
                    toDelete.push(key);
                }
            } catch (e) {
                // item has gotten messed up, just delete it as it's unusable
                toDelete.push(key);
            }
        }
        toDelete.forEach(localStorage.removeItem, localStorage);
```